### PR TITLE
Rails 5.0 -> master

### PR DESCRIPTION
**Pull request by Matt-Yorkley**
#### What? Why?

The build is now green, so this can be reviewed and tested.

**Review notes** 
- There are a few commits that fix specific problems, but the bulk of the commits are around issues with changes to the `params` object in Rails 5
- Another primary issue was around `ModelSet` classes where we were using the shovel operator (`<<`) to append directly to `ActiveRecord::Relation` objects. This doesn't work in Rails 5, and needed some adapting.

#### What should we test?

Everything! :trollface: 

#### Release notes

Upgrade to Rails 5.0

Changelog Category: User facing changes

### CODE REVIEW

**Comment by Matt-Yorkley** [comment]
There's probably a few bits I can pull into master to make this a bit smaller...

**Review by luisramos0 — COMMENTED** [review]
It's great the build is green!

There are two problems with this PR:
- I think this PR is way too big and with dual boot we have a very easy alternative to this riskier approach. I think we need to separate the raw_params changes into a PR using dual boot, it's a single if statement on the raw_params implementation, all other raw_params changes will be the same in rails 4 and rails 5. This will reduce the risk of the PR a lot.
- We need to work with the work that was done with dual boot so far, to do that there should be a commit that moves Gemfile_next.lock to Gemfile.lock. Before that I think we should merge all rails 5 PRs into master using  dual boot code, review everything, test everything (we can use dual boot to deploy to staging). When all PRs and fixes are done we can create a simple PR to update Gemfile and move Gemfile_next.lock to Gemfile.lock.

File: Gemfile [hunk 1]
```diff
@@ -4,42 +4,15 @@ source 'https://rubygems.org'
...
-  gem 'sass-rails'
-
-  gem 'db2fog'
-  gem 'unicorn'
-
-  group :test do
-    gem 'test_after_commit' # needed to test Devise callbacks
-  end
-end
+gem 'rails', '> 5.0', '< 5.1'
```
Comment (luisramos0) [inline Gemfile:7]: Is this not equivalent to ~> 5.0.0 ?

**Comment by Matt-Yorkley** [comment]
> dual boot

I guess this needs some discussion, but I think we should ditch the dual boot setup completely. Honestly it's been a _massive_ pain in the ass, and I don't think it's provided any tangible value.

File: app/assets/javascripts/templates/bulk_buy_modal.html.haml [hunk 2]
```diff
@@ -16,10 +16,7 @@
       %button.bulk-buy-add.variant-quantity{type: "button", ng: {click: "add(-1)", disabled: "!canAdd(-1)"}}>
         -# U+FF0D Fullwidth Hyphen-Minus
         －
-      %input.bulk-buy.variant-quantity{
-        type: "number",
-        min: "0",
-        max: "{{ available() }}",
+      %input.bulk-buy.variant-quantity{type: "number", min: "0", max: "{{ available() }}",
```
Comment (andrewpbrett) [inline app/assets/javascripts/templates/bulk_buy_modal.html.haml:19]: This is just formatting, right?

File: app/controllers/admin/subscriptions_controller.rb [hunk 3]
```diff
@@ -57,14 +57,14 @@ def pause
...
-      save_form_and_render
+      save_form_and_render(true, unpause: true)
     end
 
     private
 
-    def save_form_and_render(render_issues = true)
-      form = OrderManagement::Subscriptions::Form.new(@subscription, subscription_params)
+    def save_form_and_render(render_issues = true, options = {})
+      pp subscription_params
```
Comment (andrewpbrett) [inline app/controllers/admin/subscriptions_controller.rb:66]: Debugging statement?

File: app/controllers/api/base_controller.rb [hunk 4]
```diff
@@ -4,6 +4,7 @@
 
 module Api
   class BaseController < ActionController::Metal
+    include RawParams
```
Comment (andrewpbrett) [inline app/controllers/api/base_controller.rb:7]: What's the long term strategy here? Should we be looking to phase this out and use the Rails 5+ approach to param handling?

File: app/controllers/api/products_controller.rb [hunk 5]
```diff
@@ -146,7 +148,12 @@ def query_params_with_defaults
     end
 
     def product_params
-      params.require(:product).permit PermittedAttributes::Product.attributes
+      @product_params ||=
+        params.permit(product: PermittedAttributes::Product.attributes)[:product].to_h
+    end
+
+    def set_default_available_on
+      product_params[:available_on] ||= Time.zone.now
```
Comment (andrewpbrett) [inline app/controllers/api/products_controller.rb:156]: Nice 👍

File: app/controllers/spree/admin/tax_rates_controller.rb [hunk 6]
```diff
@@ -14,7 +14,7 @@ def update
       private
 
       def amount_changed?
-        BigDecimal(permitted_resource_params[:amount]) != @tax_rate.amount
+        BigDecimal(permitted_resource_params[:amount].to_s) != @tax_rate.amount
```
Comment (andrewpbrett) [inline app/controllers/spree/admin/tax_rates_controller.rb:17]: Why do we want to call `to_s` here?

File: app/services/checkout/form_data_adapter.rb [hunk 7]
```diff
@@ -6,7 +6,7 @@ class FormDataAdapter
     attr_reader :params, :shipping_method_id
 
     def initialize(params, order, current_user)
-      @params = params.dup
+      @params = params.deep_dup.to_h.with_indifferent_access
```
Comment (andrewpbrett) [inline app/services/checkout/form_data_adapter.rb:9]: This feels a little ... clunky? Brittle? Not great :) But I don't have a better suggestion, except maybe that whatever is calling this method should be expected to pass in a hash with indifferent access.

**Review by andrewpbrett — COMMENTED** [review]
Just a few comments here and there. Overall it looks like a pretty straightforward set of changes, with the vast majority having to do with params handling. My main question is whether we think we need to eventually move away from the approach of the `raw_params` concern and/or converting things to hashes with indifferent access. We don't do it in that many places, all things considered, but it feels a little hack-y when we do.

**Comment by luisramos0** [comment]
Interesting. I was able to contribute to the upgrade because of dual boot.
All the changes already in master were tested and merged. More visibility.
Easier colaboration. No long lived branches. Less risk with code tested in
master. Is this not enough?
I understand you didnt vote for dual boot in the first place. Would you
explain what is that pain?

File: app/controllers/admin/subscriptions_controller.rb (hunk 3 above)
Comment (Matt-Yorkley) [reply app/controllers/admin/subscriptions_controller.rb:66]: Yep, nice catch :see_no_evil:

**Comment by Matt-Yorkley** [comment]
> My main question is whether we think we need to eventually move away from the approach of the raw_params concern

Ideally yes, but I think we can handle it as tech-debt. There are various places where we mangle the `params` object for different reasons, and in Rails 5 it doesn't work. Some of those places would require a fair bit of refactoring to move away from, and we should definitely aim to do that and replace them with simpler usage of permitted params.

File: app/assets/javascripts/templates/bulk_buy_modal.html.haml (hunk 2 above)
Comment (Matt-Yorkley) [reply app/assets/javascripts/templates/bulk_buy_modal.html.haml:19]: Yep, Rails 5 didn't like it. It threw some fatal errors for "unbalanced brackets" when written like this. :shrug:

File: Gemfile (hunk 1 above)
Comment (Matt-Yorkley) [reply Gemfile:7]: I think this was left over from the initial step of getting bundler to actually run `bundle install`, which was pretty tricky. I'll update it now :+1:

File: app/controllers/spree/admin/tax_rates_controller.rb (hunk 6 above)
Comment (Matt-Yorkley) [reply app/controllers/spree/admin/tax_rates_controller.rb]: Updated. The `BigDecimal` class was throwing and error here unless it received a string, but I think that issue was down to how the params were being passed in the relevant spec, and not the controller itself.

File: app/services/checkout/form_data_adapter.rb (hunk 7 above)
Comment (Matt-Yorkley) [reply app/services/checkout/form_data_adapter.rb:9]: This class does a hell of a lot of mangling of the params object and expects it to work as it did in Rails 4. I think part of the issue here is that we're passing a real params object in some tests and a mock params object in other tests, and we need the object to be more-or-less uniform in any context (and it needs to behave like a Rails 4 params object). 

Yeah, it's not pretty. Actually now that I look again, I think the `#deep_dup` might be unnecessary here...

**Comment by Matt-Yorkley** [comment]
> Would you explain what is that pain?

My dev setup was totally broken by dual boot, so I couldn't actually work on the upgrade with it active. When switching back to work on PRs in Rails 4 it also regularly broke things, so (for example) I would have to comment it out every time I wanted to run migrations (I can't count the number of times I've had to do this).

> All the changes already in master were tested and merged

Those changes actually weren't tested (in Rails 5), as far as I know. Testers can't directly deploy master with Rails 5 using dual boot (without dev intervention and additional complexity). That's not the case with a simple branch and no dual-boot.

> More visibility

I don't get what visibility was really gained, or how that added actual value..?

> Easier colaboration

Hypothetically? Nobody else has actually tried working on the Rails 5 build, so there hasn't been any collaboration with dual boot. Conversely, we both collaborated on the Rails 4.1 upgrade without dual boot and it worked fine.

> No long lived branches

Even if we used dual boot, at some point we would still need a branch that switches master from Rails 4 to Rails 5, and that branch would still need to live for at least the amount of time it takes to _thoroughly test the entire app_ (lets say a week or so) and make any required adjustments/fixes. Which is the same amount of time this branch needs to live.

> Less risk with code tested in master

It hasn't been tested!

There's also the issue that _in practice_ the `Gemfile_next.lock` file hasn't been kept updated as we've been merging PRs to master for the last 2 months, and dependabot doesn't update that file either, so we're in a situation where we have two gemfiles that are completely out of sync, and because the changes have been split across 2 different files it means that we can't easily resolve those merge conflicts with **git**, but we instead have to manually merge complex diffs between different files. There's actually higher risk and complexity there (and dev time) than there would be if we could just rely on the git protocol. I've already encountered this and its a real pain. And it did create problems/errors immediately.

I think if the amount of time that was spent messing around with the dual-boot setup had been spent fixing errors in the build, we'd have merged the Rails 5 upgrade to master already.

The upgrade took about 20 hours in total and I spent at least an hour or two of that time trying to resolve problems created by dual boot.

**Comment by Matt-Yorkley** [comment]
Rebased, there's 5 or 6 less commits now :+1:

**Comment by andrewpbrett** [comment]
I marked this as ready for review so that we could see what conflicts it has with master.

**Comment by luisramos0** [comment]
Hi Matt. Ok.
I think we could argue about dual boot again and we could look at what do you mean by "My dev setup was totally broken". If you shared your problem we could find the solution. I have worked with dual boot and it worked really well  thats why i am advocating for it.
I think long lived branches are never a good solutipn.
Anyway, if you want to ditch dual boot we can just repeat the voting. Let's ask Andy, Maikel and Pau to 👍 this PR and that's it.

File: config/database.yml [hunk 8]
```diff
@@ -1,7 +1,7 @@
 defaults: &defaults
   adapter: postgresql
   encoding: unicode
-  pool: <%= ENV.fetch('OFN_DB_POOL', 5) %>
+  pool: <%= ENV.fetch('OFN_DB_POOL', 10) %>
```
Comment (sauloperez) [inline config/database.yml:4]: I checked ofn-install and we don't seem to be configuring PostgerSQL's [max_connections](https://www.postgresql.org/docs/10/runtime-config-connection.html) so I assume all servers will have the default `max_connections = 100 `. We could discuss whether or not that's a good idea though, but it falls out of this PR.

File: app/controllers/admin/enterprise_fees_controller.rb [hunk 9]
```diff
@@ -10,7 +10,7 @@ def index
 
       blank_enterprise_fee = EnterpriseFee.new
       blank_enterprise_fee.enterprise = current_enterprise
-      3.times { @collection << blank_enterprise_fee }
+      3.times { @collection.to_a << blank_enterprise_fee }
```
Comment (sauloperez) [inline app/controllers/admin/enterprise_fees_controller.rb]: 3 times?? :thinking: I don't get it.

File: Gemfile.lock [hunk 10]
```diff
@@ -526,9 +525,9 @@ GEM
     unicorn-worker-killer (0.4.4)
       get_process_mem (~> 0)
       unicorn (>= 4, < 6)
-    warden (1.2.9)
-      rack (>= 2.0.9)
-    webdrivers (4.2.0)
+    warden (1.2.7)
```
Comment (sauloperez) [inline Gemfile.lock]: that's weird. I didn't expect to see a downgrade.

File: Gemfile [hunk 11]
```diff
@@ -20,6 +20,7 @@ gem 'rails-i18n'
 gem 'rails_safe_tasks', '~> 1.0'
 
 gem "activerecord-import"
+gem "db2fog", github: "Matt-Yorkley/db2fog", branch: "rails-5"
```
Comment (sauloperez) [inline Gemfile:23]: I think it's a much better idea to fork in under openfoodfoundation's GH organization. As is, none of the other devs will have permission to touch anything on that branch if the need arises. I've had bad experiences in the past with this same issue.

File: app/controllers/admin/enterprise_fees_controller.rb [hunk 12]
```diff
@@ -10,6 +10,8 @@ def index
 
       blank_enterprise_fee = EnterpriseFee.new
       blank_enterprise_fee.enterprise = current_enterprise
+
+      @collection = @collection.to_a
       3.times { @collection << blank_enterprise_fee }
```
Comment (sauloperez) [inline app/controllers/admin/enterprise_fees_controller.rb:15]: 3 times?? :thinking: I don't get it but I guess I can live with it...

**Review by sauloperez — COMMENTED** [review]
To me, the whole point of having the dual boot was also to have the total flexibility to switch back to Rails 4.2 by just changing the env's config if anything went wrong in the rollout. Like any other feature toggle. Have we changed our mind,

File: app/controllers/admin/enterprise_fees_controller.rb (hunk 9 above)
Comment (Matt-Yorkley) [reply app/controllers/admin/enterprise_fees_controller.rb]: This is existing functionality on the enterprise fees page, we add three blank fees to the UI when the page loads so they are ready to be filled out :shrug:

File: Gemfile (hunk 11 above)
Comment (Matt-Yorkley) [reply Gemfile:23]: A yeah, I was just doing some quick tests with this to check it worked. Manually tested and it's working fine, I'll move it to OFN :+1:

File: config/database.yml (hunk 8 above)
Comment (Matt-Yorkley) [reply config/database.yml:4]: Ah I think this might have been an attempt at fixing those flaky `NetRead::Timeout` errors in the CI build, but that's now been resolved in master with a different fix. I can't quite remember though. I could drop this commit and see what the build looks like?

File: Gemfile [hunk 13]
```diff
@@ -4,38 +4,15 @@ source 'https://rubygems.org'
...
-
-  gem 'db2fog'
-  gem 'unicorn'
-
-  group :test do
-    gem 'test_after_commit' # needed to test Devise callbacks
-  end
-end
+gem 'activemerchant', '>= 1.78.0'
+gem 'angular-rails-templates', '>= 0.3.0'
```
[the remark this answers is not in the document]
Comment (Matt-Yorkley) [reply Gemfile:10]: I think the `< 1.1.0` restriction was added temporarily whilst getting through the bundle update issues, but it's currently using `1.1.0` in the Gemfile.lock and it seems to be working fine with that version.

File: Gemfile [hunk 14]
```diff
@@ -116,15 +93,12 @@ gem 'test-unit', '~> 3.4'
 gem 'coffee-rails', '~> 4.2.2'
 gem 'compass-rails'
 
-gem 'libv8', '< 8'
```
[the remark this answers is not in the document]
Comment (Matt-Yorkley) [reply Gemfile:119]: The `libv8` gem isn't a direct dependency, and it's already updated to version `8.x` in the Gemfile.lock for Rails 5.

File: config/database.yml (hunk 8 above)
Comment (sauloperez) [reply config/database.yml:4]: if this change is not required by Rails 5, I'd remove it, yes.

**Review by sauloperez — APPROVED** [review]

File: config/database.yml (hunk 8 above)
Comment (Matt-Yorkley) [reply config/database.yml]: Dropped :+1: Lets see if any phantom errors pop up in the build...

File: Gemfile [hunk 15]
```diff
@@ -4,45 +4,23 @@ source 'https://rubygems.org'
...
+gem 'sass', '<= 4.7.1'
+gem 'sass-rails', '< 6.0.0'
 
 gem 'i18n'
 gem 'i18n-js', '~> 3.8.0'
 gem 'rails-i18n'
 gem 'rails_safe_tasks', '~> 1.0'
 
 gem "activerecord-import"
+gem "db2fog", github: "openfoodfoundation/db2fog", branch: "rails-5"
```
Comment (sauloperez) [inline Gemfile:23]: IMO this needs a tech debt issue so we get to a solution that doesn't require us to maintain a fork.

Comment (Matt-Yorkley) [reply Gemfile:23]: There were some tech-debt issues related to switching to a different gem. I tried it out, but the results were not good. We lost functionality and actually the alternative gem broke some AWS-related stuff like image paths.

Comment (Matt-Yorkley) [reply Gemfile:23]: I guess we can leave an issue open to explore other options :+1:

Comment (sauloperez) [reply Gemfile:23]: we might need to explore solutions then. You know what happens when forks go unmaintained...

Comment (Matt-Yorkley) [reply Gemfile:23]: The only other option I could see was writing a custom database backup solution from scratch...

**Comment by andrewpbrett** [comment]
Moving to Test Ready unless there are objections. Probably wouldn't hurt to have @mkllnk take a look as well.

File: app/controllers/admin/variant_overrides_controller.rb [hunk 16]
```diff
@@ -103,13 +103,13 @@ def collection_errors
...
-        variant_override.permit(
+      params.permit(
+        variant_overrides: [
           :id, :variant_id, :hub_id,
           :price, :count_on_hand, :sku, :on_demand,
           :default_stock, :resettable, :tag_list
-        )
-      end
+        ]
+      ).to_h[:variant_overrides]
```
Comment (mkllnk) [inline app/controllers/admin/variant_overrides_controller.rb:112]: Can be backported.

File: engines/order_management/app/services/order_management/subscriptions/form.rb [hunk 17]
```diff
@@ -11,19 +11,21 @@ class Form
...
         @order_syncer = OrderSyncer.new(subscription)
       end
 
       def save
-        subscription.assign_attributes(subscription_params)
+        subscription.assign_attributes(subscription_params.to_h)
         return false unless valid?
 
         subscription.transaction do
+          subscription.paused_at = nil if @options[:unpause]
```
Comment (mkllnk) [inline engines/order_management/app/services/order_management/subscriptions/form.rb:28]: I think that this can be backported.

File: app/serializers/api/enterprise_shopfront_list_serializer.rb [hunk 18]
```diff
@@ -13,11 +13,11 @@ def path
 
     def icon
       icons = {
-        hub: "map_005-hub.svg",
-        hub_profile: "map_006-hub-profile.svg",
-        producer_hub: "map_005-hub.svg",
-        producer_shop: "map_003-producer-shop.svg",
-        producer: "map_001-producer-only.svg",
+        hub: "/map_005-hub.svg",
```
Comment (mkllnk) [inline app/serializers/api/enterprise_shopfront_list_serializer.rb:16]: Can this be backported? Does it fix the current group map issue?

File: app/controllers/admin/stripe_accounts_controller.rb [hunk 19]
```diff
@@ -3,7 +3,7 @@
 module Admin
   class StripeAccountsController < Spree::Admin::BaseController
     def connect
-      payload = params.slice(:enterprise_id)
+      payload = raw_params.slice(:enterprise_id)
```
Comment (mkllnk) [inline app/controllers/admin/stripe_accounts_controller.rb:6]: Isn't that what `permit` is for? Backport opportunity.

File: app/models/spree/stock_item.rb [hunk 20]
```diff
@@ -41,12 +41,12 @@ def variant
       Spree::Variant.unscoped { super }
     end
 
-    private
-
     def count_on_hand=(value)
       self[:count_on_hand] = value
     end
 
+    private
```
Comment (mkllnk) [inline app/models/spree/stock_item.rb:48]: I think that this can be backported, right?

File: app/views/spree/admin/reports/order_cycle_management.html.haml [hunk 21]
```diff
@@ -8,15 +8,15 @@
   .row
     .alpha.two.columns= label_tag nil, t(:report_customers_cycle)
     .omega.fourteen.columns
-      = f.select(:order_cycle_id_in, report_order_cycle_options(@order_cycles), {selected: params[:q][:order_cycle_id_in]}, {class: "select2 fullwidth", multiple: true})
+      = f.select(:order_cycle_id_in, report_order_cycle_options(@order_cycles), {selected: raw_params.dig(:q, :order_cycle_id_in)}, {class: "select2 fullwidth", multiple: true})
```
Comment (mkllnk) [inline app/views/spree/admin/reports/order_cycle_management.html.haml:11]: `dig` works well with `params`.
I think that this can be backported, right?

File: app/helpers/i18n_helper.rb [hunk 22]
```diff
@@ -1,6 +1,6 @@
 module I18nHelper
   def set_locale
-    UserLocaleSetter.new(spree_current_user, params[:locale], cookies).set_locale
+    UserLocaleSetter.new(spree_current_user, raw_params[:locale], cookies).set_locale
```
Comment (mkllnk) [inline app/helpers/i18n_helper.rb:3]: fetch?

File: spec/helpers/i18n_helper_spec.rb [hunk 23]
```diff
@@ -26,57 +31,57 @@
     end
 
     it "sets the chosen locale" do
-      allow(helper).to receive(:params) { { locale: "es" } }
```
Comment (mkllnk) [inline spec/helpers/i18n_helper_spec.rb:29]: Maybe we should create the method `provide_params` which takes a hash, puts it into the Parameters object and calls the `allow` part.

backportable too

File: Gemfile [hunk 24]
```diff
@@ -4,38 +4,15 @@ source 'https://rubygems.org'
 ruby "2.4.4"
 git_source(:github) { |repo_name| "https://github.com/#{repo_name}.git" }
 
-plugin 'bootboot', '~> 0.1.1' unless Bundler.settings[:frozen]
```
Comment (mkllnk) [inline Gemfile:7]: Just flagging that dualboot worked fine for me. I agree that it doesn't play well with Dependabot, manageable though. I haven't worked heaps with it yet, so can't really give a valid view on this.

File: app/helpers/enterprises_helper.rb [hunk 25]
```diff
@@ -25,7 +25,7 @@ def available_shipping_methods
   def available_payment_methods
     return [] if current_distributor.blank?
 
-    payment_methods = current_distributor.payment_methods.available(:front_end).all
+    payment_methods = current_distributor.payment_methods.available(:front_end).all.to_a
```
Comment (mkllnk) [inline app/helpers/enterprises_helper.rb:28]: Do we still need `all` if we call `to_a`?

File: app/models/concerns/line_item_stock_changes.rb [hunk 26]
```diff
@@ -0,0 +1,19 @@
...
+# no longer work correctly in relation to decrementing stock with LineItems / VariantOverrides.
+# The following methods re-instate the pre-Rails-5 versions, which work as expected.
+# https://apidock.com/rails/v4.2.9/ActiveRecord/Persistence/increment%21
+# https://apidock.com/rails/v4.2.9/ActiveRecord/Persistence/decrement%21
+
+module LineItemStockChanges
+  extend ActiveSupport::Concern
+
+  def increment!(attribute, by = 1)
+    increment(attribute, by).update_attribute(attribute, self[attribute])
```
Comment (mkllnk) [inline app/models/concerns/line_item_stock_changes.rb:13]: I think that this can be backported, right?

File: Gemfile [hunk 27]
```diff
@@ -20,6 +20,7 @@ gem 'rails-i18n'
 gem 'rails_safe_tasks', '~> 1.0'
 
 gem "activerecord-import"
+gem "db2fog", github: "openfoodfoundation/db2fog", branch: "rails-5"
```
Comment (mkllnk) [inline Gemfile:23]: Would it be worth applying the last two commits of this fork here?
https://github.com/ceresfairfood/db2fog/commits/master

We don't need all those dependencies.

**Review by mkllnk — CHANGES_REQUESTED** [review]
Awesome work, Matt! :tada: 

Really cool. It's so close. I'm tempted to push this through testing but I'm worried about the amount of commits. They are well structured and small which makes it really easy to review. Thank you. :pray: But we can reduce the risk by splitting it into two parts. I didn't comment on every commit but there are lots that can be applied on Rails 4.2. The final change to Rails 5 would become quite small. It's also an opportunity to clean up the history again because there are a few detours now.

I understand that raw_params is only a temporary solution. It's actually not making anything worse in terms of security. But I think that I saw a lot of cases in which it's just not necessary. I would try to minimise the use as much as possible. And we need a tech-debt issue for this. Using ActionController::Parameters in more tests could also help to make the specs more realistic.

What do you think? Are you open to splitting up the PR?

File: Gemfile (hunk 27 above)
Comment (Matt-Yorkley) [reply Gemfile:23]: Ah, nice. Yeah, we can do that.

**Comment by Matt-Yorkley** [comment]
Alright, I've pulled quite few commits out into #6913 and removed / tidied up various `raw_params` bits :+1:

I can rebase after #6913 is merged

**Comment by mkllnk** [comment]
This needs another rebase to resolve a conflict.

File: app/helpers/application_helper.rb [hunk 28]
```diff
@@ -1,4 +1,6 @@
 module ApplicationHelper
+  include RawParams
```
Comment (mkllnk) [inline app/helpers/application_helper.rb:2]: This looks like it should be in another commit where `raw_params` is used.

**Review by mkllnk — APPROVED** [review]
This PR still has a lot of commits, including some detours. But I don't want to delay this further. There's some great work here and it would be great to get to Rails 5. For the next upgrade, I would like to see less rushing and more smaller PRs.

I personally would like to see dualboot again. Maybe we can find a better way of doing it that works for everyone.

File: Gemfile (hunk 15 above)
Comment (sauloperez) [reply Gemfile:23]: this can be solved from ofn-install as well. Backing up a PostgreSQL it's nothing we can only do in Ruby.

**Comment by sauloperez** [comment]
There are still some code climate failures that could be addressed after this PR. Would you mind creating an issue for these @Matt-Yorkley ? I don't think we want to delay this any further.

**Comment by Matt-Yorkley** [comment]
Rebased :+1:

**Comment by filipefurtad0** [comment]
Hey @Matt-Yorkley ,

All the issues which were found are fixed.

Here are the testing notes:
https://docs.google.com/document/d/1NTBnF-yOALqK49uKo32v8wJyF5E2Guh6zdUK8I4NLxI/edit#

In addition to these, some additional test cases were done around the changes in controllers this PR introduces:

`app/controllers/admin/enterprise_fees_controller.rb`
-> tested creating fees and adding these to OC  - OK

`app/controllers/admin/enterprises_controller.rb`
`app/controllers/spree/user_passwords_controller.rb`
-> tested password setting/recovery  - OK
-> tested sign-in; create customer/enterprise/producer; change package - OK
-> tested access rights - OK
    Customer account cannot access admin;
    Accounts can only access 
-> tested enterprise permissions - OK

`app/controllers/admin/order_cycles_controller.rb`
`app/models/order_cycle.rb`
`spec/controllers/admin/order_cycles_controller_spec.rb`
-> tested CRUD operations on OCs  - OK

`app/controllers/admin/subscriptions_controller.rb`
`engines/order_management/app/services/order_management/subscriptions/form.rb`
`spec/controllers/admin/schedules_controller_spec.rb`
-> tested schedule creation - OK
-> tested subscription placement/confirmation  - OK

`app/controllers/admin/variant_overrides_controller.rb`
-> tested variant overrides  - OK

`app/controllers/api/exchange_products_controller.rb`
-> loading products from different distributors  - OK

`app/controllers/api/taxons_controller.rb`
-> tested taxon creation and attributing when creating products  - OK

`app/services/permissions/order.rb`
`app/controllers/cart_controller.rb`
`app/controllers/checkout_controller.rb`
`spec/controllers/cart_controller_spec.rb`
-> tested cart related operations, like editing, deleting, visiting  - OK

`spec/controllers/checkout_controller_concurrency_spec.rb`
-> test cart concurrency (two sessions try to snatch the last item ;-)  - OK

`app/controllers/spree/admin/products_controller.rb`
-> tested CRUD operations on products  - OK

`app/controllers/spree/admin/images_controller.rb`
-> tested image upload/display  - OK

`app/controllers/spree/admin/payment_methods_controller.rb`
-> tested payments/payment capture: cash/paypal - OK

`app/controllers/spree/admin/reports_controller.rb`
`engines/order_management/spec/services/order_management/reports/enterprise_fee_summary/renderers/csv_renderer_spec.rb`
-> tested reports  - OK

`app/controllers/spree/admin/search_controller.rb`
-> tested searching for customers (when placing BO orders)  - OK

`app/controllers/spree/admin/tax_rates_controller.rb`
-> tested setting inclusive taxes  - OK

`app/controllers/spree/admin/variants_controller.rb`
-> tested searching for variants when (when placing BO orders)  - OK

I think this is good to go to be merged :tada:

**Comment by andrewpbrett** [comment]
@Matt-Yorkley the gemfile needs another conflict resolution
